readme.md
# Progress

Scriptable macOS application to show and control progress indicators

<img src="preview.png" alt="Preview of a scriptable progress indicator">

## Installation

A code-signed and notarized build of this project can be downloaded from [here](https://adriannier.de/downloads/Progress_1.0.3.56.zip).

## Usage with AppleScript

Run one of the demonstration scripts in the `scripts` directory of this repository for an example on how to create and control progress indicators. 

Basic syntax is:

```
tell application "Progress"
    set myIndicator to make new indicator with properties {visible:true, title:"Processing…"}
end tell

repeat with i from 1 to 100
    
    -- Do something
    
    tell application "Progress" to tell myIndicator
        set percentage to i
    end tell
    
end repeat

tell application "Progress" to tell myIndicator
    set title to "Done"
end tell

```
When the `percentage` is set to a value below zero an indeterminate progress is displayed. 

The `indicator` is automatically set to be complete when the `percentage` reaches 100. No further `percentage` updates except with `-1` are allowed past this point.

If there are no updates to a newly created `indicator` for 15 seconds, the user is presented with a cancel button. 

If the cancel button is pressed your script has another 15 seconds to finish processing otherwise the user will be able to close the `indicator`.

## Usage without AppleScript

If you would like to show an indicator without AppleScript then start the application from the command line or in a shell script with two arguments each pointing to a `.json` file.

**First argument**: Path to data file

Example:
```
{
    "percentage": -1,
    "title": "This is the title",
    "message": "This is a descriptive message",
    "icon": "/System/Library/CoreServices/Dock.app/Contents/Resources/url@2x.png"
}
```
As long as the user does not cancel the progress, this file is checked every second for changes to be mirrored by the indicator.

**Second argument**: Path to `cancel-file`

Example
```
#!/bin/bash

~/Downloads/Progress.app/Contents/MacOS/Progress ~/Downloads/swift-progress/json/progress.json ~/Downloads/swift-progress/_cancel.json

```

Once the user presses the cancel button the `cancel-file` is created to give your application or script a chance to know of the user’s intent. At this point you are encouraged to delete the `cancel-file` which will update the indicator accordingly. If the `cancel-file` does not disappear within 15 seconds, the user is allowed to close the progress indicator and the `cancel-file` is deleted.

## Building

This project was successfully build with Xcode 11.3.1 and Swift 5 on macOS 10.14.6.


changelog.md
# Changes

## 1.0.3

- Minor UI optimizations (cancel button position and font weights)

## 1.0.2

- Setting `percentage` to 100.0 (`real`) instead of 100 (`integer`) with AppleScript now correctly sets the indicator as done
- Fixed a bug that caused a newly created indicator to never time out and thus never showed a cancel button
- Rewrote `demo.applescript` and `demo_error.applescript` by encapsulating indicator communication in discrete functions with the benefit that the script will still run even if Progress is not installed. 

## 1.0.1

- First public release


swift-progress/Scripting.swift
//
//  Scripting.swift
//  swift-progress
//
//  Created by Adrian Nier on 2020-05-27.
//  Copyright © 2020 Adrian Nier. All rights reserved.
//

import Cocoa

extension AppDelegate {
    
    func application(_ sender: NSApplication, delegateHandlesKey key: String) -> Bool {
        return key == "indicators"
    }
    
    func insertValue(_ object: Indicator, inIndicatorsAt index: Int) {
        
        indicators.append(object)
        
        if let indicator = indicators.last {
            indicator.show()
        }
        
    }
    
    func removeObjectFromIndicatorsAtIndex(_ index: Int) {
        indicators.remove(at: index)
    }
    
    
}

extension Indicator {
    
    override var objectSpecifier: NSScriptObjectSpecifier {
        
        let appDescription = NSApplication.shared.classDescription as! NSScriptClassDescription
        
        let specifier = NSUniqueIDSpecifier(
            containerClassDescription: appDescription,
            containerSpecifier: nil,
            key: "indicators",
            uniqueID: id
        )
        
        return specifier
        
    }
    
    @objc override var scriptingProperties: [String : Any]? {
        
        // This happens when properties are read or set as a whole
        
        get { return currentState.asDict() }
        set { update(newValue) }
        
    }
    
    @objc func show(_ command: NSScriptCommand?) {
        
        update(IndicatorState(isVisible: true))
        
    }

    @objc func hide(_ command: NSScriptCommand?) {
        
        update(IndicatorState(isVisible: false))
        
    }
    
    @objc func close(_ command: NSScriptCommand?) {
        
        stopCancelTimer()
        update(IndicatorState(isVisible: false))
        appDelegate.removeIndicator(self)
    }
    
    @objc func complete(_ command: NSScriptCommand?) {
        
        update(IndicatorState(completed: true))
        
    }
    
    @objc func abort(_ command: NSScriptCommand?) {
        
        update(IndicatorState(aborted: true))
        
    }
    
    @objc func cancel(_ command: NSScriptCommand?) {
        
        cancelButtonAction()
        
    }

    
}


swift-progress/utilities.swift
//
//  utilities.swift
//

import Foundation

func debugLog(_ msg: String) {
   
    if debugMode { print(msg) }
    
}

func errorLog(_ msg: String) {
    
    print("Error: \(msg)")
    
}

func infoLog(_ msg: String) {
    
    print(msg)
    
}

func parentDirectory(at: URL) -> URL {

    return at.deletingLastPathComponent()

}

func parentDirectory(at: String) -> String {
    
    return parentDirectory(at: URL(fileURLWithPath: at)).path

}

func fileExists(at: URL) -> Bool {
    return fileExists(at: at.path)
}

func fileExists(at: String) -> Bool {
    
    var isDirectory: ObjCBool = false
    if !FileManager.default.fileExists(atPath: at, isDirectory: &isDirectory) || isDirectory.boolValue {
        return false
    } else {
        return true
    }
    
}

func directoryExists(at: URL) -> Bool {
    return directoryExists(at: at.path)
}

func directoryExists(at: String) -> Bool {
    
    var isDirectory: ObjCBool = false
    if FileManager.default.fileExists(atPath: at, isDirectory: &isDirectory) && isDirectory.boolValue {
        return true
    } else {
        return false
    }
    
}

func isReadable(at: URL) -> Bool {
    return isReadable(at: at.path)
}

func isReadable(at: String) -> Bool {
    return FileManager.default.isReadableFile(atPath: at)
}

func isWritable(at: URL) -> Bool {
    return isWritable(at: at.path)
}

func isWritable(at: String) -> Bool {
    return FileManager.default.isWritableFile(atPath: at)
}


swift-progress/AppDelegate.swift
//
//  AppDelegate.swift
//  swift-progress
//
//  Created by Adrian Nier on 2020-05-25.
//  Copyright © 2020 Adrian Nier. All rights reserved.
//

import Cocoa

let debugMode = false

@NSApplicationMain
@objcMembers
class AppDelegate: NSObject, NSApplicationDelegate {
    
    var removedIndicators: [Indicator] = []
    @objc var indicators: [Indicator] = []
    
    var exitTimer = Timer()
    
    // MARK: -
    // MARK: JSON
    // Properties that deal with the fact that a JSON file has been specified as command line argument
    
    var jsonIndicator: Indicator?
    var jsonPath: String?
    var jsonCancelPath: String?
    
    var jsonTimer = Timer()
    let jsonFileUpdateFrequency = 1.0
    
    // MARK: -
    // MARK: Application
    
    func applicationDidFinishLaunching(_ notification: Notification) {
        
        initalizeArguments()
        
        if jsonPath != nil {
            
            cancelJsonTimer()
                
            if let state = json_stateFromFile() {
                
                jsonIndicator = Indicator(state: state)
                
                if jsonIndicator!.completed {
                    
                    infoLog("Timer already complete when the JSON file was loaded")
                    
                    removeIndicator(jsonIndicator!)
                    
                } else {
                    
                    jsonIndicator!.update(IndicatorState(isVisible: true))
                    
                    jsonTimer = Timer.scheduledTimer(
                        timeInterval: jsonFileUpdateFrequency,
                        target: self,
                        selector: #selector(self.json_performScheduledStateUpdate),
                        userInfo: nil,
                        repeats: true
                    )
                    
                }
                
            } else {
                
                errorLog("Could not instantiate indicator based on file information")
                terminateWithNoIndicators()
                
            }
            
        }
        
    }
    
    func terminateWithNoIndicators() {
        
        if self.indicators.count == 0 && self.jsonIndicator == nil {
            
            infoLog("Will exit in 10 seconds if there are still no indicators")
            
            exitTimer = Timer.scheduledTimer(
                timeInterval: 10,
                target: self,
                selector: #selector(self.performScheduledExit),
                userInfo: nil,
                repeats: true
            )
            
        }
        
    }
    
    func cancelExitTimer() {
        
        if exitTimer.isValid { exitTimer.invalidate() }
        
    }
    
    @objc func performScheduledExit() {
        
        DispatchQueue.global(qos: .userInteractive).async {
            
            if self.indicators.count == 0 && self.jsonIndicator == nil {
                
                DispatchQueue.main.async {
                    
                    debugLog("Exiting with no indicators found")
                    NSApplication.shared.terminate(self)
                    
                }
                
            } else {
                
                DispatchQueue.main.async {
                    
                    debugLog("Canceling exit as there are still indicators")
                    self.cancelExitTimer()
                    
                }
                
            }
            
        }
        
    }
    
    // MARK: -
    // MARK: Manage indicators
    
    func removeIndicator(_ indicator: Indicator) {
        
        // FIXME: Why does the application crash if we don't keep a reference to the indicator?
        removedIndicators.append(indicator)
        
        indicator.window.delegate = nil
        indicator.window.isReleasedWhenClosed = true
        indicator.window.close()
        
        if jsonIndicator != nil && indicator.id == jsonIndicator!.id {
            
            cancelJsonTimer()
            jsonIndicator = nil
            
        } else {
            
            for (index, thisIndicator) in indicators.enumerated() {
                
                if thisIndicator.id == indicator.id {
                    indicators.remove(at: index)
                    break
                }
                
            }
            
        }
        
        terminateWithNoIndicators()
        
    }
    
    // MARK: -
    // MARK: JSON file
    
    
    func isJsonBasedIndicator(_ indicator: Indicator) -> Bool {
        
        if jsonIndicator != nil {
            
            if jsonIndicator!.id == indicator.id {
                
                return true
            }
            
        }
        
        return false
        
    }
    
    func cancelJsonTimer() {
        
        if jsonTimer.isValid { jsonTimer.invalidate() }
        
    }
    
    @objc func json_performScheduledStateUpdate() {
        
        debugLog("Trying to update state from JSON file")
        
        DispatchQueue.global(qos: .userInteractive).async {
            
            if let newInfo = self.json_stateFromFile() {
                
                DispatchQueue.main.async {
                    
                    if let indicator = self.jsonIndicator {
                        
                        indicator.update(newInfo)
                        
                    }
                    
                }
                
            }
            
        }
        
    }
    
    func json_stateFromFile() -> IndicatorState? {
        
        do {
            
            if jsonPath != nil {
                if let jsonData = try json_readStateFile(jsonPath!) {
                    if let info = try json_parseStateData(jsonData) {
                        return info
                    }
                }
            }
            
        } catch {
            
            debugLog("\(error)")
            
        }
        
        return nil
        
    }
    
    func json_readStateFile(_ path: String) throws -> Data? {
        
        if let jsonData = try String(contentsOfFile: path).data(using: .utf8) {
            return jsonData
        }
        
        return nil
    }
    
    func json_parseStateData(_ jsonData: Data) throws -> IndicatorState? {
        
        return try JSONDecoder().decode(IndicatorState.self, from: jsonData)
        
    }
    
    func json_writeCancelFile() -> Bool {
        
        if jsonCancelPath != nil {
            
            let jsonCancelTime = Date()
            
            do {
                
                let jsonData = try JSONEncoder().encode(["cancelation_time": jsonCancelTime])
                
                if let jsonString = String(data: jsonData, encoding: .utf8) {
                    
                    try jsonString.write(toFile: jsonCancelPath!, atomically: true, encoding: .utf8)
                    
                    return true
                }
                
            } catch {
                
                errorLog("Failed to write cancelation file: \(error)")
                
            }
            
        }
        
        return false
    }
    
    
    
}


swift-progress/CommandLine.swift
//
//  CommandLine.swift
//  swift-progress
//
//  Created by Adrian Nier on 2020-05-27.
//  Copyright © 2020 Adrian Nier. All rights reserved.
//

import Cocoa

extension AppDelegate {
        
    func initalizeArguments() {
        
        if CommandLine.arguments.count > 1 {
            
            // At least one command line argument
            
            let specifiedInfoPath = NSString(string: CommandLine.arguments[1]).expandingTildeInPath
            
            if specifiedInfoPath != "-NSDocumentRevisionsDebugMode" {
                
                
                if fileExists(at: specifiedInfoPath) {
                    if isReadable(at: specifiedInfoPath) {
                        jsonPath = specifiedInfoPath
                    } else {
                        errorLog("The JSON file is not readable at \(specifiedInfoPath)")
                    }
                } else {
                    errorLog("The JSON file is missing at \(specifiedInfoPath)")
                }
                
                if jsonPath != nil && CommandLine.arguments.count > 2 {
                    
                    // At least two command line arguments
                    
                    let specifiedCancelPath = NSString(string: CommandLine.arguments[2]).expandingTildeInPath
                    
                    if !fileExists(at: specifiedCancelPath) || isWritable(at: specifiedCancelPath) {
                        
                        let parent = parentDirectory(at: specifiedCancelPath)
                        
                        if isWritable(at: parent) {
                            
                            let fm = FileManager.default
                            
                            if fm.isDeletableFile(atPath: specifiedCancelPath) {
                                
                                jsonCancelPath = specifiedCancelPath
                                
                                if fileExists(at: jsonCancelPath!) {
                                    
                                    do {
                                        try fm.removeItem(atPath: jsonCancelPath!)
                                    } catch {
                                        debugLog("Could not delete cancel file at \(jsonCancelPath!): \(error)")
                                    }
                                    
                                }
                                
                            } else {
                                
                                debugLog("The cancel file is not deletable at \(specifiedInfoPath)")
                                
                            }
                            
                        } else {
                            debugLog("The directory for the cancel file is not writable at \(parent)")
                        }
                        
                    } else {
                        
                        debugLog("The cancel file is missing at \(specifiedInfoPath)")
                        
                    }
                    
                }
                
            }
            
        } else {
            debugLog("No .json file specified")
        }
        
    }
    
}


swift-progress/IndicatorState.swift
//
//  ProgressInfo.swift
//  swift-progress
//
//  Created by Adrian Nier on 2020-05-25.
//  Copyright © 2020 Adrian Nier. All rights reserved.
//

import Foundation

func defaultIndicatorState() -> IndicatorState {
    
    return IndicatorState(
        title: "",
        message: "",
        percentage: -2.0,
        icon: "",
        completed: false,
        aborted: false,
        canceled: false,
        isVisible: false
    )
    
}

func modifiedIndicatorState(currentState: IndicatorState, newState: IndicatorState) -> IndicatorState {
    
    var modifiedState = currentState
    
    if newState.title != nil { modifiedState.setValueForKey("title", value: newState.title) }
    if newState.message != nil { modifiedState.setValueForKey("message", value: newState.message) }
    if newState.percentage != nil { modifiedState.setValueForKey("percentage", value: newState.percentage) }
    if newState.icon != nil { modifiedState.setValueForKey("icon", value: newState.icon) }
    if newState.completed != nil { modifiedState.setValueForKey("completed", value: newState.completed) }
    if newState.aborted != nil { modifiedState.setValueForKey("aborted", value: newState.aborted) }
    if newState.canceled != nil { modifiedState.setValueForKey("canceled", value: newState.canceled) }
    if newState.isVisible != nil { modifiedState.setValueForKey("isVisible", value: newState.isVisible) }
    
    return modifiedState
    
}

struct IndicatorState: Codable {
    
    var title: String?
    var message: String?
    var percentage: Double?
    var icon: String?
    var completed: Bool?
    var aborted: Bool?
    var canceled: Bool?
    var isVisible: Bool?
    
    mutating func setValueForKey(_ key: String, value: Any??) {
        
        if value != nil {
            
            switch key {
                
                case "title": title = value as? String
                case "message":  message = value as? String
                case "percentage":  percentage = value as? Double
                case "icon": icon = value as? String
                case "completed": completed = value as? Bool
                case "aborted": aborted = value as? Bool
                case "canceled": canceled = value as? Bool
                case "isVisible": isVisible = value as? Bool
                
            default:
                debugLog("IndicatorState: Unknown key \"\(key)\".")
                
            }
    
        }
    
    }
    
    func asDict() -> [String: Any] {
        
        let defaultState = defaultIndicatorState()
        
        var dict: [String: Any] = [:]
        
        dict["title"] = title ?? defaultState.title
        dict["message"] = message ?? defaultState.message
        dict["percentage"] = percentage ?? defaultState.percentage
        dict["icon"] = icon ?? defaultState.icon
        dict["completed"] = completed ?? defaultState.completed
        dict["aborted"] = aborted ?? defaultState.aborted
        dict["canceled"] = canceled ?? defaultState.canceled
        dict["isVisible"] = isVisible ?? defaultState.isVisible
        
        return dict
        
    }
    
    func valueForKey(_ key: String) -> Any {
        
        if let value = asDict()[key] {
            return value
        } else {
            debugLog("IndicatorState: Unknown key \"\(key)\".")
            return ""
        }
        
    }
    
}
